### PR TITLE
Fix VirtualizedList skip measure items - Improve scroll experience

### PR DESCRIPTION
**Pull request by LucioChavezFuentes**
After investigation and feedback I have a new proposal of pull requests to fix the Inverted VirtualizedList in web. Some of the root issues are in react-native-web's end, another issue is fixed updating [VirtualizedList from react-native](https://github.com/facebook/react-native/tree/main/packages/virtualized-lists/Lists):

- **Problem 1**: VirtualizedList's code relies on that `onLayout` measures items without considering transformations **—>** [Problem Explanation and Solution in this PR](https://github.com/necolas/react-native-web/pull/2501).

- **Problem 2**: Exists inaccurate measures in onLayout when the setTimeout is executed and any of the targeted nodes (node or relativeNode) for measure are unmounted. **—>** [Problem Explanation and Solution in this PR (same PR that solves problem 1)](https://github.com/necolas/react-native-web/pull/2501).

- **Problem 3**: The scroll position can be forced to areas where items have not been measured and the list skips measuring items if the user scrolls too fast. **—>** **Problem Explanation and Solution below.**

---

### Problem 3 ###

Even if we measure correctly the items of VirtualizedList we still have a problem with the scroll experience if the user scrolls too fast.

The react-native's [VirtualizedList updated code](https://github.com/facebook/react-native/tree/main/packages/virtualized-lists/Lists) fixes this problem but the [react-native-web's wheel patch](https://github.com/necolas/react-native-web/blob/bf24fb5a3aef1b836235533221722d0218ece01a/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L745-L747) for inverted VirtualizedList opens the possibility for users to force scroll any part of the list and not give enough time for the list to measure items to map their position.

**Solution**

1. Update VirtualizedList from react-native (included in this PR) 
2. Limit the scroll position based on VirtalizedList's totalCellLength (the sum of heights of all items measured)

Screenshot 2023-03-22 at 17 20 29

With all solutions applied and updating VirtualizedList from react-native we have a solid VirtualizedList in web, inverted or non-inverted.

Thank you for reading @necolas, if you have any questions or concerns please let me know how I can help.

### CODE REVIEW

**Review by janicduplessis — APPROVED** [review]
Looks good. I wonder if we would we be able to vendor https://github.com/facebook/react-native/tree/main/packages/virtualized-lists package directly instead?

**Comment by LucioChavezFuentes** [comment]
> Looks good. I wonder if we would we be able to vendor https://github.com/facebook/react-native/tree/main/packages/virtualized-lists package directly instead?

I think we can, I will give ve a try. To be honest I still learning how the files are organized and exported in RN vs RNW.

**Comment by necolas** [comment]
I appreciate all the hard work you've put into identifying and fixes issues in the RN lists.

But I would like to stop forking FlatList/VirtualizedList, which is why I've been trying to get those modules packaged out of RN and imported into RNW. Then all patches can be made in once place (the RN repo) which should simplify making fixes and help cross-platform apps with consistency. So I'm unlikely to merge this PR at this time. I understand it's probably also a challenge to get fixes to the lists merged in the RN repo at the moment (sorry about that), but having only one place to get changes reviewed and merged should at least make things better going forward.

Let me know if there's anything else I can do

**Comment by LucioChavezFuentes** [comment]
Thank you for review @necolas but I still need to understand the workflow between RN and RNW. 
1. If I want to make changes to VirtualizedList I should apply a PR to to RN's VirtualizedList. And RNW pulls the changes from RN? 
2. As for the RNW's patches  (like the [inverted wheel](https://github.com/necolas/react-native-web/blob/bf24fb5a3aef1b836235533221722d0218ece01a/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L745-L747)) they are exclusive of RNW repo? If that is true then the PR goes to RNW?

This PR is only pulling changes from RN and making a change to a RNW patch.

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,10 @@ build
 coverage
 dist
 node_modules
+
+# Visual studio
+.vscode
+.vs
+
+# OS X
+.DS_Store
```
Comment (necolas) [inline .gitignore:12]: Please remove these. You can add these to your local machine's git-ignore

File: packages/react-native-web/package.json [hunk 2]
```diff
@@ -26,7 +26,9 @@
     "create-react-class": "^15.7.0",
     "fbjs": "^3.0.4",
     "inline-style-prefixer": "^6.0.1",
+    "memoize-one": "^6.0.0",
     "normalize-css-color": "^1.0.2",
+    "nullthrows": "^1.1.1",
```
Comment (necolas) [inline packages/react-native-web/package.json:31]: I didn't see where this gets used

**Review by necolas — COMMENTED** [review]
This needs a rebase please

**Comment by necolas** [comment]
I'm also open to including this patch for VirtualizedList. Let me know if it looks good and is something Expensify wants https://github.com/necolas/react-native-web/pull/2436

Then all the current patches related to scroll wheel will be included at least, and porting anything over to RN might be easier.

File: packages/react-native-web/package.json (hunk 2 above)
Comment (LucioChavezFuentes) [reply packages/react-native-web/package.json:31]: [Here](https://github.com/necolas/react-native-web/blob/928637f7548af04e57e2a6378398171284e7f7f5/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L1597), inside VirtualizedList/index.js

**Comment by LucioChavezFuentes** [comment]
@necolas this PR also has been rebased!

**Comment by LucioChavezFuentes** [comment]
Should I move the files to a `virtualized-lists` folder? Like this [comment](https://github.com/necolas/react-native-web/pull/2502#pullrequestreview-1357228029) suggests?

**Comment by necolas** [comment]
The flow tests are currently failing

**Comment by LucioChavezFuentes** [comment]
> The flow tests are currently failing

Im on it, looks like there a lot of errors.

**Comment by LucioChavezFuentes** [comment]
@necolas I have a problem with the modules flow type:
Screenshot 2023-03-31 at 18 29 13

Could you help me how to tackle this issue, please?

**Comment by LucioChavezFuentes** [comment]
This [commit](https://github.com/necolas/react-native-web/commit/3a024ee308e06857307b9cf80da1517c5f860c13#diff-994373d173205075304ad456ea564044ea1ed2e22ebcce284efb65494a24d870) breaks any RNW app that uses `react-native-gesture-handler` and `@react-navigation/native` libraries. Had to revert the commit locally in order to test my PR in Expensify App. Should it be an issue in react-native-web or in the libraries?

**Comment by LucioChavezFuentes** [comment]
> @necolas I have a problem with the modules flow type: Screenshot 2023-03-31 at 18 29 13
> 
> Could you help me how to tackle this issue, please?


@necolas I added [these files](https://github.com/necolas/react-native-web/pull/2502/files#diff-07b5d114dbce9438ad5c80e2aff55b9663c83c72d9f0bf5b275e9906c6c028edR2) to fix the modules types problems. Is this ok?

**Comment by necolas** [comment]
Hmm, do you know why React Native doesn't have these errors despite not including these module declarations?

**Comment by LucioChavezFuentes** [comment]
> Hmm, do you know why React Native doesn't have these errors despite not including these module declarations?

No, It's what I've been trying to figure out.

**Comment by LucioChavezFuentes** [comment]
@necolas I fixed the typing module from react native perspective. React-Native [declares the types in `.flowconfig`](https://github.com/facebook/react-native/blob/main/.flowconfig#L24). Also, It doesn't ignore the node_modules, except for specific modules.

This PR fixes https://github.com/necolas/react-native-web/issues/2432

**Comment by LucioChavezFuentes** [comment]
Hello, @necolas. This PR is ready, it fixes https://github.com/necolas/react-native-web/issues/2432
Let me know if can help with something else. I fixed the typing modules based on the react-native configuration.

**Comment by LucioChavezFuentes** [comment]
Thank you @necolas, hope I was helpful.

**Comment by necolas** [comment]
Thanks! This patch is in release 0.19.2

**Pull request closed without merging**